### ISSUE TITLE
Unresolved symbol errors when linking libprotobuf v3.5.2 dynamically on Windows

### ISSUE THREAD

**Issue by nehaljwani** [google/protobuf#4717]
```
yyyy.obj : error LNK2001: unresolved external symbol "class google::protobuf::internal::ExplicitlyConstructed<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > google::protobuf::internal::fixed_address_empty_string" (?fixed_address_empty_string@internal@protobuf@google@@3V?$ExplicitlyConstructed@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@123@A)
xxxx.obj : error LNK2001: unresolved external symbol "public: static void * __cdecl google::protobuf::internal::GenericTypeHandler<class google::protobuf::Message>::GetMaybeArenaPointer(class google::protobuf::Message *)" (?GetMaybeArenaPointer@?$GenericTypeHandler@VMessage@protobuf@google@@@internal@protobuf@google@@SAPEAXPEAVMessage@34@@Z)
xxxx.obj : error LNK2001: unresolved external symbol "public: static class google::protobuf::Arena * __cdecl google::protobuf::internal::GenericTypeHandler<class google::protobuf::Message>::GetArena(class google::protobuf::Message *)" (?GetArena@?$GenericTypeHandler@VMessage@protobuf@google@@@internal@protobuf@google@@SAPEAVArena@34@PEAVMessage@34@@Z)
xxxx.obj : error LNK2001: unresolved external symbol "public: static class google::protobuf::Message * __cdecl google::protobuf::internal::GenericTypeHandler<class google::protobuf::Message>::NewFromPrototype(class google::protobuf::Message const *,class google::protobuf::Arena *)" (?NewFromPrototype@?$GenericTypeHandler@VMessage@protobuf@google@@@internal@protobuf@google@@SAPEAVMessage@34@PEBV534@PEAVArena@34@@Z)
build\lib.win-amd64-3.6\_xxxx.cp36-win_amd64.pyd : fatal error LNK1120: 4 unresolved externals
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\link.exe' failed with exit status 1120
```
The one related to `ExplicitlyConstructed` was initially fixed in #3216 and then broken again with the 'merge' at 09354db1434859a31a3c81abebcc4018d42f2715 (released) and then fixed again in 0400cca3236de1ca303af38bf81eab332d042b7c (unreleased, only in v3.6.0 release candidates)

I am not sure what's the correct solution for the remaining errors. Probably this?:
```diff
diff --git a/src/google/protobuf/repeated_field.h b/src/google/protobuf/repeated_field.h
index b47ea99..c3b0acc 100644
--- a/src/google/protobuf/repeated_field.h
+++ b/src/google/protobuf/repeated_field.h
@@ -633,7 +633,7 @@ class LIBPROTOBUF_EXPORT RepeatedPtrFieldBase {
 };
 
 template <typename GenericType>
-class GenericTypeHandler {
+class LIBPROTOBUF_EXPORT GenericTypeHandler {
  public:
   typedef GenericType Type;
   typedef GenericType WeakType;
```

**Comment by acozzette** [comment]
@nehaljwani Sorry about all the trouble with this. I am going to merge 3.6.x into the master branch and it sounds like that may fix some of the errors. Do you want to send out a pull request with your other fix?

**Comment by nehaljwani** [comment]
@acozzette Sure, I can do that.

**Comment by kprakhya0321** [comment]
I am getting the following error:

error LNK2019: unresolved external symbol "class google::protobuf::internal::ExplicitlyConstructed<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > google::protobuf::internal::fixed_address_empty_string" (?fixed_address_empty_string@internal@protobuf@google@@3V?$ExplicitlyConstructed@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@123@A) referenced in function "public: __cdecl leonardo::data::config::Setting::Setting(class leonardo::data::config::Setting const &)" (??0Setting@config@data@leonardo@@QEAA@AEBV0123@@Z)

I have noticed this issue with several versions of protocolbuffers/protobuf up to v3.10.0. I am currently using v3.1.0 and v3.3.0, where I have tried the suggested fix #2407 but that doesn't work. I used the earlier versions to get my code that uses protocol buffers to compile quicker. I noticed the compilation was much slower with v3.9.2 and v3.10.0. More specifically, I tried to append LIBPROTOBUF_EXPORT everywhere that #2407 has suggested. I am using Visual Studio 2017 and cmake. Any thoughts?
